phpBB/includes/acp/acp_inactive.php

```diff
@@ -207,37 +197,37 @@ function main($id, $mode)
 					if ($row = $db->sql_fetchrow($result))
 					{
 						// Send the messages
-						if (!class_exists('messenger'))
-						{
-							include($phpbb_root_path . 'includes/functions_messenger.' . $phpEx);
-						}
-
-						$messenger = new messenger();
 						$usernames = $user_ids = array();
 
+						/** @var \phpbb\di\service_collection */
+						$messenger = $phpbb_container->get('messenger.method_collection');
+						$messenger_collection_iterator = $messenger->getIterator();
+
 						do
 						{
-							$messenger->template('user_remind_inactive', $row['user_lang']);
-
-							$messenger->set_addresses($row);
-
-							$messenger->anti_abuse_headers($config, $user);
-
-							$messenger->assign_vars(array(
-								'USERNAME'		=> html_entity_decode($row['username'], ENT_COMPAT),
-								'REGISTER_DATE'	=> $user->format_date($row['user_regdate'], false, true),
-								'U_ACTIVATE'	=> generate_board_url() . "/ucp.$phpEx?mode=activate&u=" . $row['user_id'] . '&k=' . $row['user_actkey'])
-							);
-
-							$messenger->send($row['user_notify_type']);
+							foreach ($messenger_collection_iterator as $messenger_method)
+							{
+								if ($messenger_method->get_id() == $user_row['user_notify_type'] || $user_row['user_notify_type'] == $messenger_method::NOTIFY_BOTH)
+								{
+									$messenger_method->template('user_remind_inactive', $row['user_lang']);
+									$messenger_method->set_addresses($row);
+									$messenger_method->anti_abuse_headers($config, $user);
+									$messenger_method->assign_vars([
+										'USERNAME'		=> html_entity_decode($row['username'], ENT_COMPAT),
+										'REGISTER_DATE'	=> $user->format_date($row['user_regdate'], false, true),
+										'U_ACTIVATE'	=> generate_board_url() . "/ucp.$phpEx?mode=activate&u=" . $row['user_id'] . '&k=' . $row['user_actkey'],
+									]);
+
+									$messenger_method->send();
+									$messenger_method->save_queue();
```

Comment:
It seems very inefficient to have the save queue call inside the loops here instead of after the do while loop.